### PR TITLE
compiler: add CONFIG_COMPILER_TRACK_MACRO_EXPANSION

### PR DESCRIPTION
**Pull request by gmarull**
Add a new compiler option to control tracking locations of tokens across macro
expansions. While the compiler diagnostics will be reduced when disabled, it can be
useful in situations where the expansion chain is long enough to _hide_
useful error messages.

Updated DT guide: when debugging devicetree related issues, it may be useful to turn macro
expansion tracking off. Let's document this in the troubleshooting
section.

Example:

```diff
diff --git a/drivers/regulator/regulator_npm6001.c b/drivers/regulator/regulator_npm6001.c
index 32e8c3192e..e9881608d6 100644
--- a/drivers/regulator/regulator_npm6001.c
+++ b/drivers/regulator/regulator_npm6001.c
@@ -714,7 +714,7 @@ static const struct regulator_driver_api api = {
                          NPM6001_BUCKMODEPADCONF_BUCKMODE1PULLD_ENABLED) |    \
                         (DT_INST_PROP(inst, nordic_buck_mode2_pull_down) *    \
                          NPM6001_BUCKMODEPADCONF_BUCKMODE2PULLD_ENABLED)),    \
-               .pad_val = ((DT_INST_PROP(inst, nordic_ready_high_drive) *     \
+               .pad_val = ((DT_INST_PROP(inst, nordic_ready_high_driv) *     \
                             NPM6001_PADDRIVESTRENGTH_READY_HIGH) |            \
                            (DT_INST_PROP(inst, nordic_nint_high_drive) *      \
                             NPM6001_PADDRIVESTRENGTH_NINT_HIGH) |             \
```

default setup:
```
west build -b nrf52840dk_nrf52840 samples/shields/npm6001_ek -p

...

FAILED: zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj 
ccache /home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DNRF52840_XXAA -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/gmarull/ws/nordic/zephyrproject/zephyr/include -I/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/include/generated -I/home/gmarull/ws/nordic/zephyrproject/zephyr/s
oc/arm/nordic_nrf/nrf52 -I/home/gmarull/ws/nordic/zephyrproject/zephyr/lib/posix/getopt/. -I/home/gmarull/ws/nordic/zephyrproject/zephyr/soc/arm/nordic_nrf/common/. -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/nordic/nrfx -I/home/gmarull/ws/nordic/zephyrproj
ect/modules/hal/nordic/nrfx/drivers/include -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/nordic/nrfx/mdk -I/home/gmarull/ws/nordic/zephyrproject/zephyr/modules/hal_nordic/nrfx/. -I/home/gmarull/ws/nordic/zephyrproject/modules/debug/segger/SEGGER -I/home/gmarull/ws/nordic/zephyrproject/modules/debug/segger/Config -I/home/gmarull/ws/no
rdic/zephyrproject/zephyr/modules/segger/. -isystem /home/gmarull/ws/nordic/zephyrproject/zephyr/lib/libc/minimal/include -isystem /home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/include -isystem /home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/include-fixed -fno-str
ict-aliasing -Os -imacros /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs --sysroot=/home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/gmarull/ws/nordic/zephyrproject/zephy
r/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home
/gmarull/ws/nordic/zephyrproject/zephyr/samples/shields/npm6001_ek=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/gmarull/ws/nordic/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/gmarull/ws/nordic/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -D_POSIX_THREADS -std=c99 -nostdinc -MD -MT zephyr/drivers/regulator/CMakeFil
es/drivers__regulator.dir/regulator_npm6001.c.obj -MF zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj.d -o zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj -c /home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c
In file included from /home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util.h:17,                                                                                                                                                                                                                                                      
                 from /home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:21,
                 from /home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/device.h:12,                      
                 from /home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/drivers/i2c.h:25,                                                                                                                                                                                                                                                   
                 from /home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c:10:
/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_generated.h:15321:46: error: 'DT_N_S_soc_S_i2c_40003000_S_pmic_70_S_regulators_P_nordic_ready_high_driv' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_i2c_40003000_S_pmic_70_S_regulators_P_nordic_ready_high_drive'?
15321 | #define DT_N_INST_0_nordic_npm6001_regulator DT_N_S_soc_S_i2c_40003000_S_pmic_70_S_regulators
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:72:26: note: in definition of macro '__DEBRACKET'
   72 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:64:9: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   64 |         __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |         ^~~~~~~~~~~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:59:9: note: in expansion of macro '__COND_CODE'
   59 |         __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:180:9: note: in expansion of macro 'Z_COND_CODE_1'
  180 |         Z_COND_CODE_1(_flag, _if_1_code, _else_code)
      |         ^~~~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:3906:9: note: in expansion of macro 'COND_CODE_1'
 3906 |         COND_CODE_1(DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT),   \
      |         ^~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:609:32: note: in expansion of macro 'DT_CAT3'
  609 | #define DT_PROP(node_id, prop) DT_CAT3(node_id, _P_, prop)
      |                                ^~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:3345:34: note: in expansion of macro 'DT_PROP'
 3345 | #define DT_INST_PROP(inst, prop) DT_PROP(DT_DRV_INST(inst), prop)
      |                                  ^~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_N_INST_0_nordic_npm6001_regulator'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:104:26: note: in expansion of macro 'UTIL_PRIMITIVE_CAT'
  104 | #define UTIL_CAT(a, ...) UTIL_PRIMITIVE_CAT(a, __VA_ARGS__)
      |                          ^~~~~~~~~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:328:31: note: in expansion of macro 'UTIL_CAT'
  328 | #define DT_INST(inst, compat) UTIL_CAT(DT_N_INST, DT_DASH(inst, compat))
      |                               ^~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:3154:27: note: in expansion of macro 'DT_INST'
 3154 | #define DT_DRV_INST(inst) DT_INST(inst, DT_DRV_COMPAT)
      |                           ^~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/devicetree.h:3345:42: note: in expansion of macro 'DT_DRV_INST'
 3345 | #define DT_INST_PROP(inst, prop) DT_PROP(DT_DRV_INST(inst), prop)
      |                                          ^~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c:717:30: note: in expansion of macro 'DT_INST_PROP'
  717 |                 .pad_val = ((DT_INST_PROP(inst, nordic_ready_high_driv) *     \
      |                              ^~~~~~~~~~~~
/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_generated.h:17631:59: note: in expansion of macro 'REGULATOR_NPM6001_DEFINE_ALL'
17631 | #define DT_FOREACH_OKAY_INST_nordic_npm6001_regulator(fn) fn(0)
      |                                                           ^~
/home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/sys/util_internal.h:105:36: note: in expansion of macro 'DT_FOREACH_OKAY_INST_nordic_npm6001_regulator'
  105 | #define UTIL_PRIMITIVE_CAT(a, ...) a##__VA_ARGS__
      |                                    ^
/home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c:743:1: note: in expansion of macro 'DT_INST_FOREACH_STATUS_OKAY'
  743 | DT_INST_FOREACH_STATUS_OKAY(REGULATOR_NPM6001_DEFINE_ALL)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
[148/197] Building C object modules/hal_nordic/nrfx/CMakeFiles/modules__hal_nordic__nrfx.dir/home/gmarull/ws/nordic/zephyrproject/modules/hal/nordic/nrfx/drivers/src/nrfx_twi.c.obj
ninja: build stopped: subcommand failed.

```

without macro expansion tracking:
```
west build -b nrf52840dk_nrf52840 samples/shields/npm6001_ek -p -- -DCONFIG_COMPILER_NO_TRACK_MACRO_EXPANSION=y

...

FAILED: zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj 
ccache /home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DKERNEL -DNRF52840_XXAA -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I/home/gmarull/ws/nordic/zephyrproject/zephyr/include -I/home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/include/generated -I/home/gmarull/ws/nordic/zephyrproject/zephyr/soc/arm/nordic_nrf/nrf52 -I/home/gmarull/ws/nordic/zephyrproject/zephyr/lib/posix/getopt/. -I/home/gmarull/ws/nordic/zephyrproject/zephyr/soc/arm/nordic_nrf/common/. -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/cmsis/CMSIS/Core/Include -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/nordic/nrfx -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/nordic/nrfx/drivers/include -I/home/gmarull/ws/nordic/zephyrproject/modules/hal/nordic/nrfx/mdk -I/home/gmarull/ws/nordic/zephyrproject/zephyr/modules/hal_nordic/nrfx/. -I/home/gmarull/ws/nordic/zephyrproject/modules/debug/segger/SEGGER -I/home/gmarull/ws/nordic/zephyrproject/modules/debug/segger/Config -I/home/gmarull/ws/nordic/zephyrproject/zephyr/modules/segger/. -isystem /home/gmarull/ws/nordic/zephyrproject/zephyr/lib/libc/minimal/include -isystem /home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/include -isystem /home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/include-fixed -fno-strict-aliasing -Os -imacros /home/gmarull/ws/nordic/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -ftrack-macro-expansion=0 -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs --sysroot=/home/gmarull/zephyr-sdk-0.16.0/arm-zephyr-eabi/arm-zephyr-eabi -imacros /home/gmarull/ws/nordic/zephyrproject/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/gmarull/ws/nordic/zephyrproject/zephyr/samples/shields/npm6001_ek=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/gmarull/ws/nordic/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/gmarull/ws/nordic/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -D_POSIX_THREADS -std=c99 -nostdinc -MD -MT zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj -MF zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj.d -o zephyr/drivers/regulator/CMakeFiles/drivers__regulator.dir/regulator_npm6001.c.obj -c /home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c
/home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c:743:1: error: 'DT_N_S_soc_S_i2c_40003000_S_pmic_70_S_regulators_P_nordic_ready_high_driv' undeclared here (not in a function); did you mean 'DT_N_S_soc_S_i2c_40003000_S_pmic_70_S_regulators_P_nordic_ready_high_drive'?
  743 | DT_INST_FOREACH_STATUS_OKAY(REGULATOR_NPM6001_DEFINE_ALL)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      | DT_N_S_soc_S_i2c_40003000_S_pmic_70_S_regulators_P_nordic_ready_high_drive
[178/197] Building C object zephyr/kernel/CMakeFiles/kernel.dir/sched.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /home/gmarull/ws/nordic/zephyrproject/zephyr/build
```

### CODE REVIEW

**Review by fabiobaltieri — DISMISSED** [review]
Cool!

**Comment by nordicjm** [comment]
Nice addition but not so sure on what it displays, the original output is a mess but it does include this valuable line:
```
/home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c:717:30: note: in expansion of macro 'DT_INST_PROP'
  717 |                 .pad_val = ((DT_INST_PROP(inst, nordic_ready_high_driv) *     \
```
Whereas in the new output it just points to:
```
  743 | DT_INST_FOREACH_STATUS_OKAY(REGULATOR_NPM6001_DEFINE_ALL)
```
And although that's the root macro where it starts from, I'm not sure that's going to help someone that doesn't know much about device tree (etc.) to find the issue

**Review by iuliana-prodan — DISMISSED** [review]

**Comment by erwango** [comment]
> And although that's the root macro where it starts from, I'm not sure that's going to help someone that doesn't know much about device tree (etc.) to find the issue

Agree it isn't perfect. Anyway, since it is not enabled by default and even if correctly documented, it will remain an option which won't be commonly used.

**Review by erwango — DISMISSED** [review]

**Comment by gmarull** [comment]
> Nice addition but not so sure on what it displays, the original output is a mess but it does include this valuable line:
> 
> ```
> /home/gmarull/ws/nordic/zephyrproject/zephyr/drivers/regulator/regulator_npm6001.c:717:30: note: in expansion of macro 'DT_INST_PROP'
>   717 |                 .pad_val = ((DT_INST_PROP(inst, nordic_ready_high_driv) *     \
> ```
> 
> Whereas in the new output it just points to:
> 
> ```
>   743 | DT_INST_FOREACH_STATUS_OKAY(REGULATOR_NPM6001_DEFINE_ALL)
> ```
> 
> And although that's the root macro where it starts from, I'm not sure that's going to help someone that doesn't know much about device tree (etc.) to find the issue

Yeah, compiler _always_ outputs some useful message, the problem is that with expansion it quickly gets buried, so you need to scroll up and find it.

**Review by nordicjm — DISMISSED** [review]

**Review by iuliana-prodan — DISMISSED** [review]

**Review by erwango — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

File: Kconfig.zephyr [hunk 1]
```diff
@@ -378,6 +378,14 @@ config COMPILER_SAVE_TEMPS
 	  Instruct the compiler to save the temporary intermediate files
 	  permanently. These can be useful for troubleshooting build issues.
 
+config COMPILER_NO_TRACK_MACRO_EXPANSION
+	bool "Do not track macro expansion"
+	help
+	  Do not track locations of tokens across macro expansions. While the
+	  compiler diagnostics will be reduced, it can be useful in situations
+	  where the expansion chain is long enough to _hide_ useful error
+	  messages.
```
Comment (stephanosio) [inline Kconfig.zephyr:387]: It is arguably more intuitive to make this `COMPILER_TRACK_MACRO_EXPANSION` and `default y`.

**Review by stephanosio — DISMISSED** [review]

File: Kconfig.zephyr (hunk 1 above)
Comment (erwango) [reply Kconfig.zephyr:387]: I vote to change as well

**Review by fabiobaltieri — APPROVED** [review]

File: Kconfig.zephyr (hunk 1 above)
Comment (gmarull) [reply Kconfig.zephyr]: done!

**Review by stephanosio — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]
Awesome. Thank you.

**Review by iuliana-prodan — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
Really nice, thanks!

**Comment by moose-rivieh** [comment]
Thank you, @gmarull. This is helpful.

**Pull request merged**